### PR TITLE
Allow integers to be used as map keys

### PR DESCRIPTION
**Pull request by bjorn3**

### CODE REVIEW

File: json/src/ser.rs [hunk 1]
```diff
@@ -489,44 +489,44 @@ impl<'a, W, F> ser::Serializer for MapKeySerializer<'a, W, F>
         Err(Error::Syntax(ErrorCode::KeyMustBeAString, 0, 0))
     }
 
-    fn serialize_isize(&mut self, _value: isize) -> Result<()> {
-        Err(Error::Syntax(ErrorCode::KeyMustBeAString, 0, 0))
+    fn serialize_isize(&mut self, value: isize) -> Result<()> {
+        self.ser.serialize_str(&value.to_string())
```
Comment (oli-obk) [inline json/src/ser.rs]: That allocates on the heap, can you find a way around that? Like write a `"`, the serialize normally, write another quote?

**Comment by bjorn3** [comment]
Added a test and it doesn't heap allocate anymore.

**Comment by dtolnay** [comment]
> This has probably been discussed before... I'm not sure we want this, and if we do, we need deserialization along with it

We do want this. The discussion is in #45. Do you have any reasons against?

File: json_tests/tests/test_json.rs [hunk 2]
```diff
@@ -1501,10 +1501,24 @@ fn test_deserialize_from_stream() {
...
     let map = treemap!(
-        1 => 2,
-        3 => 4
+        true => 2,
+        false => 4
+    );
+
+    match serde_json::to_vec(&map).unwrap_err() {
+        serde_json::Error::Syntax(serde_json::ErrorCode::KeyMustBeAString, 0, 0) => {}
+        _ => panic!("integers used as keys"),
```
Comment (dtolnay) [inline json_tests/tests/test_json.rs]: Please update this error message

File: json_tests/tests/test_json.rs [hunk 3]
```diff
@@ -1501,10 +1501,24 @@ fn test_deserialize_from_stream() {
...
+    }
+}
+
+#[test]
+fn test_serialize_rejects_adt_keys() {
```
Comment (dtolnay) [inline json_tests/tests/test_json.rs]: Please update the panic message below - it still refers to integers

File: json_tests/tests/test_json.rs [hunk 4]
```diff
@@ -1723,3 +1737,26 @@ fn test_stack_overflow() {
...
+#[test]
+fn test_allow_ser_integers_as_map_keys(){
+    let map = treemap!(
+        1 => 2,
+        2 => 4,
+        -1 => 6,
+        -2 => 8
+    );
+    
+    serde_json::to_vec(&map).unwrap();
```
Comment (dtolnay) [inline json_tests/tests/test_json.rs]: Please test the actual JSON string that it produces. You might be surprised what it is right now:

```json
{"\""-2"\"":8,"\""-1"\"":6,"\""1"\"":2,"\""2"\"":4}
```

File: json_tests/tests/test_json.rs [hunk 5]
```diff
@@ -1723,3 +1737,26 @@ fn test_stack_overflow() {
...
+}
+
+/*#[test]
+fn test_allow_de_integers_as_map_keys(){
```
Comment (dtolnay) [inline json_tests/tests/test_json.rs]: Are you interested in implementing this yourself in this PR or in another PR, or would you prefer to leave it to somebody else?

File: json/src/ser.rs [hunk 6]
```diff
@@ -489,44 +489,48 @@ impl<'a, W, F> ser::Serializer for MapKeySerializer<'a, W, F>
...
 
-    fn serialize_i32(&mut self, _value: i32) -> Result<()> {
-        Err(Error::Syntax(ErrorCode::KeyMustBeAString, 0, 0))
+    fn serialize_i32(&mut self, value: i32) -> Result<()> {
+        self.serialize_i64(value as i64)
     }
 
-    fn serialize_i64(&mut self, _value: i64) -> Result<()> {
-        Err(Error::Syntax(ErrorCode::KeyMustBeAString, 0, 0))
+    fn serialize_i64(&mut self, value: i64) -> Result<()> {
```
Comment (dtolnay) [inline json/src/ser.rs]: I think this should be:

```rust
try!(self.ser.writer.write_all(b"\""));
try!(self.ser.serialize_i64(value));
try!(self.ser.writer.write_all(b"\""));
Ok(())
```

The `serialize_str` method writes a double-quoted JSON string. We just need to write a byte directly into the output.

**Comment by oli-obk** [comment]
> Do you have any reasons against?

No, I want this, but I didn't remember that issue or what became of the discussion. Let's do it

File: json/src/ser.rs (hunk 6 above)
Comment (bjorn3) [inline json/src/ser.rs]: Done

File: json_tests/tests/test_json.rs (hunk 5 above)
Comment (bjorn3) [inline json_tests/tests/test_json.rs]: Will try it in another PR.

File: json_tests/tests/test_json.rs (hunk 2 above)
Comment (oli-obk) [inline json_tests/tests/test_json.rs]: You missed thismin your update

File: json_tests/tests/test_json.rs (hunk 3 above)
Comment (oli-obk) [inline json_tests/tests/test_json.rs]: Same here

File: json_tests/tests/test_json.rs (hunk 4 above)
Comment (oli-obk) [inline json_tests/tests/test_json.rs]: And here

**Comment by bjorn3** [comment]
Pushed a commit to improve build times on travis. Feel free to omit it.

**Comment by oli-obk** [comment]
Hmm github doesn't allow me to approve @dtolnay's requested changes

Everything looks good to me now. Also, good idea with the cargo cache. I wonder if that will go into crazy overdrive someday with incremental compilation.

File: json_tests/tests/test_json.rs (hunk 5 above)
Comment (bjorn3) [inline json_tests/tests/test_json.rs]: Created PR serde-rs/serde#655

**Comment by arthurprs** [comment]
Did we regress this between 0.9 and 1.0? I'm still investigating but I have some code choking on deserializing a struct containing a map<int,...>

**Comment by dtolnay** [comment]
Yes, I'm working on a fix.

**Pull request merged**